bootstore/src/node.rs

```diff
@@ -0,0 +1,117 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! An individual bootstore node.
+//!
+//! Most logic is contained here, but networking sits on top.
+//! This allows easier testing of clusters and failure situations.
+
+use slog::Logger;
+use slog::{info, o};
+use sprockets_host::Ed25519Certificate;
+use std::net::SocketAddrV6;
+use std::sync::Arc;
+use std::sync::Mutex;
+use uuid::Uuid;
+
+use crate::db::Db;
+use crate::messages::*;
+use crate::trust_quorum::SerializableShareDistribution;
+
+/// An error returned by a Node
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Version {0} messages are unsupported.")]
+    UnsupportedVersion(u32),
```

Comment:
Why isn't this a case in `NodeError` instead?

Reply:
My initial thinking was that printable errors like this would end up getting converted to `NodeErrors` which themselves are part of a `NodeOpResult`, which itself is part of a `NodeResponse`. But that is problematic because we then lose the correlation `id`. 

Really, the underlying issue is that I want to be able to return errors with the `?` operator from the handle methods. But I can do that by just returning a `NodeError`. The top level code just has to know how package that up into an eventual `NodeResponse`.  Maybe `handle` should just return a `NodeResponse` no matter what. The only downside there is that tests have to match to get the error out of the response, but that seems reasonable? I can also just create a `dispatch` method called by handle that returns a `result`, and then all `handle` does is package that into a `NodeResponse`. That seems like potentially overkill though.

Reply:
What if you put a full-fledged `Result` into `NodeResponse`? Something like

```rust
pub struct NodeResponse {
    pub version: u32,
    pub id: u64,
    pub result: Result<NodeOpResult, NodeError>,
}
```

and then remove the `Error` case from `NodeOpResult`?

Reply:
I changed the code to return `Result<NodeOpResult, NodeError>` from the handler methods, and changed `handle` itself to just return `NodeResponse`. This works, its just a touch weird because we end up packaging a `NodeError` into `NodeOpResult` inside `handle`. But all in all I prefer it to having every method have to do that packaging and not be able to use the `?` operator.

Reply:
Fixed in [07e043e](https://github.com/oxidecomputer/omicron/pull/1656/commits/07e043e547b44d9b299d3014407ea1f85ad457b0)

Thanks!